bin/e2hs-writer/src/reader.rs

```diff
@@ -0,0 +1,384 @@
+            beacon_block_proof,
+            beacon_block_root: block.tree_hash_root(),
+            slot,
+            execution_block_proof,
+        };
+
+        let header_with_proof = HeaderWithProof {
+            header,
+            proof: BlockHeaderProof::HistoricalSummaries(proof),
+        };
+        let withdrawals: Vec<Withdrawal> =
+            payload.withdrawals.iter().map(Withdrawal::from).collect();
+        let body = BlockBody(AlloyBlockBody {
+            transactions,
+            ommers: vec![],
+            withdrawals: Some(Withdrawals::new(withdrawals)),
+        });
+        let receipts = self
+            .execution_api
+            .get_era_receipts(
+                block_number,
+                payload.transactions.len() as u64,
+                payload.receipts_root,
+            )
+            .await?;
+        Ok(AllBlockData {
+            block_number,
+            header_with_proof,
+            body,
+            receipts,
+        })
+    }
+
+    pub fn iter_blocks(self) -> impl Stream<Item = anyhow::Result<AllBlockData>> {
+        let mut current_block = self.starting_block;
+        stream! {
+        while current_block < self.ending_block {
+            if current_block < MERGE_BLOCK_NUMBER {
+                yield self.get_pre_merge_block_data(current_block);
+            } else if current_block < SHANGHAI_BLOCK_NUMBER {
+                yield self.get_pre_capella_block_data(current_block).await;
+            } else if current_block < CANCUN_BLOCK_NUMBER {
+                yield self.get_pre_deneb_block_data(current_block).await;
+            } else {
+                yield Err(anyhow!("Invalid block number: {current_block}"));
+            }
+            current_block += 1;
+        }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ethportal_api::types::{
+        consensus::{
+            beacon_block::{BeaconBlockBellatrix, BeaconBlockCapella},
+            beacon_state::BeaconState,
+            fork::ForkName,
+        },
+        execution::header_with_proof::{
+            build_historical_roots_proof, build_historical_summaries_proof,
+            BlockProofHistoricalRoots, BlockProofHistoricalSummaries,
+        },
+    };
+    use serde_yaml::Value;
+    use ssz::Decode;
+
+    use super::*;
+
+    #[rstest::rstest]
+    #[case(15539558, 4702208, 575, "block_proofs_bellatrix/beacon_block_proof-15539558-cdf9ed89b0c43cda17398dc4da9cfc505e5ccd19f7c39e3b43474180f1051e01.yaml")]
+    #[case(15547621, 4710400, 576, "block_proofs_bellatrix/beacon_block_proof-15547621-96a9313cd506e32893d46c82358569ad242bb32786bd5487833e0f77767aec2a.yaml")]
+    #[case(15555729, 4718592, 577, "block_proofs_bellatrix/beacon_block_proof-15555729-c6fd396d54f61c6d0f1dd3653f81267b0378e9a0d638a229b24586d8fd0bc499.yaml")]
+    #[tokio::test]
+    async fn test_pre_capella_proof_generation(
+        #[case] block_number: u64,
+        #[case] slot: u64,
+        #[case] _epoch: u64,
```

Comment:
```suggestion
```

if `_epoch` isn't used do we need it?

Reply:
I just thought it might be useful information to leave there for future reference, so people don't need to go digging to figure out what epoch we're operating on. But, i've updated it to just a comment